### PR TITLE
E1681. Refactor and test the quizzing feature + E1633. Refactor different question types from quiz feature

### PR DESCRIPTION
**Pull request by qizhongzhi**

### CODE REVIEW

**Comment by LeoJY** [comment]
@yangsong8 please merge this pull request instead of the E1633.
This one has already include the code of E1633.
Thanks

**Comment by yangsong8** [comment]
Just want to make sure, the bug I told Lei yesterday has been fixed in this pull request as well, right?

**Comment by qizhongzhi** [comment]
Yes, the bug has been fixed

**Comment by yangsong8** [comment]
Okay, I will test this on my local machine then.

**Comment by zhoilettest** [comment]
@yangsong8 Just a reminder, this pull request comes from zqi3 branch instead of master branch of our repository. The commit hash is 7de6a46.

**Comment by yangsong8** [comment]
The logic of the questionnaires_controller can be improved further.

When a new quiz is created, the method 'valid_quiz' will be called, then, the questions and questionnaire will be created if the input is valid by calling 'create_questionnaire'. But when user updates an existing quiz, the method 'update_quiz' will be called. If you look into 'update_quiz', you may find that the quiz questions/choices are not validated before got saved. 

The issue this will cause is that, user may create a valid quiz, then edit the quiz and make the quiz not valid.

E.g.:
image

File: app/controllers/questionnaires_controller.rb [hunk 1]
```diff
@@ -373,48 +373,31 @@ def valid_quiz
...
-        elsif type == 'TF' # TF is not disabled. We need to test TF later.
-          if params[:new_choices][i.to_s]["TF"].nil?
-            # A correct option isnt selected for a true/false question
-            valid = "Please select a correct answer for all questions"
+        choice_info = params[:new_choices][i.to_s][type]    #choice info for one question of its type 
+        if choice_info == nil
+          valid = "Please select a correct answer for all questions"
+          break
+        else
+          valid = @new_question.isvalid(choice_info)
```
Comment (yangsong8) [inline app/controllers/questionnaires_controller.rb]: method name should be 'DYaihao.com '

**Comment by qizhongzhi** [comment]
That is an issue. However, the project description just told us to refactor the valid_quiz method so we may miss some logic of other methods in the questionnaire_controller.  We will demo tomorrow and we need to make a video for the project so we may not have time to fix it now.

File: app/models/multiple_choice_checkbox.rb [hunk 2]
```diff
@@ -1,10 +1,88 @@
...
+    choice_info.each do |idx, value|
+      if value[:txt] == ''
+        valid = "Please make sure every question has text for all options"
+        break
+      end
+      if value[:iscorrect] == 1.to_s
+        correct_count+=1
+      end
+    end
+    if correct_count == 0
```
Comment (yangsong8) [inline app/models/multiple_choice_checkbox.rb]: if it equals 1, there should be another warning message "You only specified one correct answer for a multiple-choice checkbox question. "

**Comment by yangsong8** [comment]
Okay, I understand, please go ahead and make the video base on what you have achieved. You guys have already made enough progress.

**Comment by qizhongzhi** [comment]
Thank you!

File: app/models/multiple_choice_checkbox.rb [hunk 3]
```diff
@@ -1,10 +1,88 @@
...
+      html += 'value="' + quiz_question_choices[i].txt + '" size="40" />'
+
+      html += '</td></tr>'
+    end
+
+    html.html_safe
   end
 
-  def complete(count, answer = nil)
+  def complete
```
Comment (yangsong8) [inline app/models/multiple_choice_checkbox.rb]: The complete method should display the question, then the choices.

**Comment by yangsong8** [comment]
Minor changes. See commit 86f1e6aab7d9e1eb43396f32a93108ab7cdc7ab9 and d15998f3a57d55a1319c56e64263c4a5afc01e31

**Pull request merged**